mlir/lib/Dialect/MIOpen/Generator/AmdArchDb.cpp

```diff
@@ -0,0 +1,39 @@
+//===- AmdArchDb.cpp - Dtabase of AMD GPU features ------------------===//
+//
+// Part of the MLIR Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/MIOpen/Generator/AmdArchDb.h"
+
+#include "mlir/Dialect/MIOpen/MIOpen.h"
+
+#include "llvm/ADT/StringSwitch.h"
+
+using namespace mlir;
+using namespace mlir::miopen;
+
+static constexpr AmdArchInfo gcnFeatures(GemmFeatures::none),
+    cdnaFeatures(GemmFeatures::xdlops | GemmFeatures::dot |
+                 GemmFeatures::atomic_add),
+    rdnaNoDotFeatures(GemmFeatures::wave32),
+    rdnaFeatures(GemmFeatures::wave32 | GemmFeatures::dot),
+    gfx11Features(GemmFeatures::wave32 | GemmFeatures::dot |
+                  GemmFeatures::atomic_add);
+
+AmdArchInfo mlir::miopen::lookupArchInfo(StringRef chip) {
+  return llvm::StringSwitch<AmdArchInfo>(chip)
+      .Cases("gfx908", "gfx90a", "gfx940", cdnaFeatures)
+      // gfx906 has the dot product instructions, uniquely
+      .Case("gfx906", AmdArchInfo(GemmFeatures::dot))
+      .StartsWith("gfx9", gcnFeatures)
+      .Cases("gfx1011", "gfx1013", rdnaNoDotFeatures)
+      .Cases("gfx1010", "gfx1012", rdnaFeatures)
+      // All gfx103x are the same for us
+      .StartsWith("gfx103", rdnaFeatures)
```

Comment:
Still not a fan of this string matching.. for gfx103* it will strcmp 10 times for every lookup.

Perhaps we should have a SmallDenseMap that can contains a "gfx<major>" that can be used when an exact match doesn't exist.

Reply:
String matches hang around the compiler a lot, I would argue this isn't particularly unreasonable. And I just looked at the disassembly - this does not at all go to anything resembling a linear series of `strcmp` calls. Lots of `xor` and the like.

Reply:
That is, I don't think the complexity of trying to write out this out even more explicitly is worth it, especially when `-O3` is very good at its job.

Reply:
Sure, not surprised they have a full lexer-like parser/decoder. But this list will grow over time and the complexity will grow with it.

Maybe should use <major><minor> lookup, and where no <minor> matches, using the default <major>.

Reply:
Check latest comments on the ticket.

Reply:
(Ok, I'll give you checking major and minor version separately to be safe)